### ISSUE TITLE
[Feat]: Appium 3 🚀 

### ISSUE THREAD

**Issue by KazuCocoa** [appium/appium#20792]
Appium 3 will keep the same architecture as Appium v2, but we want to make changes such as dropping old NodeJS versions/bumping up Express version to 5 etc. It means this major versionup will not be a big major version update like Appium v1 to v2. Each driver such as UIA2/XCUITest etc will keep their responsibility in their repositories, so their functionalities won't be affected by this change.

Working branch: https://github.com/appium/appium/tree/appium3

## Items

- Support NodeJS 20+ (npm 10+)
    - NodeJS v18 will be EOL in April 2025
- Express v5
- Switch `node.http` to `node.http2` for Express, get rid of the third-party spdy dependency
- Drop driver-specific handling from the main core
    - zip/ipa, any other specific extensions in download helpers (base-driver)
    - and others if we find out
- drop old deprecated non-w3c endpoints
- change `/sessions` endpoint https://github.com/appium/appium/issues/19856
    - available behind `allow-insecure`
    - can add more data if needed(?)
- Investigate possibility of moving to ESM
- Better https://github.com/appium/appium-inspector integration
- Make security flags driver-specific to prevent unwanted reuse by other drivers/plugins

## target RC
- April 2025 (by following NodeJS v18 EOL)

**Comment by KazuCocoa** [comment]
@jlipps @mykola-mokhnach @eglitise Feel free to modify the main description, linking issues tec

**Comment by KazuCocoa** [comment]
Just a note:

We got an idea about appium-inspector.

- making appium inspector available via `appium` server (without appium-inspector electron app)
    - Then, perhaps making https://github.com/projectxyzio/appium-inspector 's web edition an appium plugin is a good idea...? Then, our setup command could install the plugin as well

**Comment by saikrishna321** [comment]
Follow up note: 
 
* making appium inspector available via /inspector. We will continue to release electron app as well.

**Comment by KazuCocoa** [comment]
Btw, the appium-inspector as part of appium server idea can do as a plugin without Appium 3 like with https://github.com/appium/appium/blob/master/packages/fake-plugin/lib/plugin.js#L63-L69. It makes sense to add the endpoint as a plugin since both modules have different release cycles. The appium server could add a shortcut as part of Appium 2 major version.
One challenge here is probably how to modify package.json in https://github.com/appium/appium-inspector to make it work as a plugin as well. https://github.com/appium/appium-inspector/blob/main/package.json#L35 is already ready to build for browser usage, so return the index.html as `/inspector` endpoint response as a plugin will make `/inspector` endpoint work like https://inspector.appiumpro.com/

**Comment by KazuCocoa** [comment]
I made a new issue https://github.com/appium/appium/issues/20834 about the inspector thing since this can be done out of appium 3 work as an existig plugin.

**Comment by Koushikvij** [comment]
Does Appium3, address the element identification issue in react Native apps when the app has stacks that are > 62 levels of depth?

**Comment by eglitise** [comment]
> Does Appium3, address the element identification issue in react Native apps when the app has stacks that are > 62 levels of depth?

No, because that issue is on Apple's end - please refer to the discussion in #14825. Even if it was fixable from Appium side, the fix would be applied in the XCUITest driver, and not the platform-independent core Appium server.
